File: Semester 6/LP2/A3/A3.py
import sys, heapq
from collections import defaultdict
from math import inf
from rich import print


def selectionSort(A):
    U = A.copy()
    for i in range(len(A)):
        min_idx = i
        for j in range(i+1, len(A)):
            if A[min_idx] > A[j]:
                min_idx = j

        A[i], A[min_idx] = A[min_idx], A[i]

    print(f'Selection Sort:\nUnsorted array: {U}\nSorted array: {A}')


def jobScheduling(arr):
    n = len(arr)
    arr.sort(key=lambda x: x[1])
    result = []
    maxHeap = []

    for i in range(n - 1, -1, -1):
        if i == 0:
            slots_available = arr[i][1]
        else:
            slots_available = arr[i][1] - arr[i - 1][1]

        heapq.heappush(maxHeap, (-arr[i][2], arr[i][1], arr[i][0]))

        while slots_available and maxHeap:
            profit, deadline, job_id = heapq.heappop(maxHeap)
            slots_available -= 1
            result.append([job_id, deadline])

    result.sort(key=lambda x: x[1])
    print(f'\n\nJob Scheduling Problem:\nFollowing is maximum profit sequence of jobs: {result}')


class PGraph:
    def __init__(self, vertices, graph):
        self.V = vertices
        self.graph = graph

    def minKey(self, key, mstSet):
        min = sys.maxsize

        for v in range(self.V):
            if key[v] < min and mstSet[v] == False:
                min = key[v]
                min_index = v

        return min_index

    def primMST(self):
        key = [sys.maxsize] * self.V
        parent = [None] * self.V
        key[0] = 0
        mstSet = [False] * self.V

        parent[0] = -1

        for cout in range(self.V):
            u = self.minKey(key, mstSet)

            mstSet[u] = True
            for v in range(self.V):
                if self.graph[u][v] > 0 and mstSet[v] == False and key[v] > self.graph[u][v]:
                        key[v] = self.graph[u][v]
                        parent[v] = u

        print(f'\n\nPrim’s Minimum Spanning Tree:\nEdge \tWeight')
        minimumCost = 0
        for i in range(1, self.V):
            print(f'{parent[i]} -- {i} == {self.graph[i][parent[i]]}')
            minimumCost += self.graph[i][parent[i]]
        print(f'Minimum cost = {minimumCost}')


class KGraph:
    def __init__(self, vertices, graph):
        self.V = vertices
        self.graph = []

        for i in range(self.V):
            for j in range(i, self.V):
                if graph[i][j] != 0:
                    self.graph.append([i, j, graph[i][j]])

    def find(self, parent, i):
        if parent[i] == i:
            return i
        return self.find(parent, parent[i])

    def union(self, parent, rank, x, y):
        xroot = self.find(parent, x)
        yroot = self.find(parent, y)

        if rank[xroot] < rank[yroot]:
            parent[xroot] = yroot
        elif rank[xroot] > rank[yroot]:
            parent[yroot] = xroot
        else:
            parent[yroot] = xroot
            rank[xroot] += 1

    def KruskalMST(self):
        result = []
        i = 0
        e = 0
        self.graph = sorted(self.graph, key = lambda item: item[2])

        parent = []
        rank = []

        for node in range(self.V):
            parent.append(node)
            rank.append(0)

        while e < self.V - 1:
            u, v, w = self.graph[i]
            i = i + 1
            x = self.find(parent, u)
            y = self.find(parent, v)

            if x != y:
                e = e + 1
                result.append([u, v, w])
                self.union(parent, rank, x, y)

        minimumCost = 0
        print(f'\n\nKruskal’s Minimum Spanning Tree:\nEdge \tWeight')
        for u, v, weight in result:
            minimumCost += weight
            print(f'{u} -- {v} == {weight}')
        print(f'Minimum cost = {minimumCost}')


def dijkstra(inp: list, source: str, dest: str) -> int:
    graph = defaultdict(list)
    weight = {}
    path = [dest]

    for i in range(len(inp)):
        for j in range(i, len(inp)):
            if inp[i][j] != 0:
                s, d, c = i, j, inp[i][j]
                graph[s].append(d)
                weight[f'{s} {d}'] = int(c)

                graph[d].append(s)
                weight[f'{d} {s}'] = int(c)

    Q = list(graph.keys())
    A, d, p = [], {}, defaultdict(list)

    for v in Q:
        d[v] = inf
        p[v] = []

    d[source] = 0

    while Q:
        u = min(Q, key=lambda x: d[x])
        A.append(u)
        Q.remove(u)

        for v in set(graph[u]).intersection(Q):
            alt = d[u] + weight[f'{u} {v}']
            if d[v] > alt:
                d[v] = alt
                p[v].append(u)

        if u == dest:
            break

    key = dest
    while p[key]:
        path.append(p[key][-1])
        key = p[key][-1]
    path.reverse()

    print(f'\n\nDijkstra Single-Source Shortest Path:\nPath: {path}\nMinimum Cost: {d[dest]}\n\n')

if __name__ == '__main__':

    A = [64, 25, 12, 22, 11]
    selectionSort(A)

    A = [
        ['A', 2, 100],
        ['B', 1, 19],
        ['C', 2, 27],
        ['D', 1, 25],
        ['E', 3, 15]
    ]
    jobScheduling(A)

    graph = [
        [0, 4, 0, 0, 0, 0, 0, 8, 0],
        [4, 0, 8, 0, 0, 0, 0,11, 0],
        [0, 8, 0, 7, 0, 4, 0, 0, 2],
        [0, 0, 7, 0, 9,14, 0, 0, 0],
        [0, 0, 0, 9, 0,10, 0, 0, 0],
        [0, 0, 4,14,10, 0, 2, 0, 0],
        [0, 0, 0, 0, 0, 2, 0, 1, 6],
        [8,11, 0, 0, 0, 0, 1, 0, 7],
        [0, 0, 2, 0, 0, 0, 6, 7, 0]
    ]

    g = PGraph(9, graph)
    g.primMST()

    g = KGraph(9, graph)
    g.KruskalMST()

    dijkstra(graph, 0, 4)

File: Semester 6/LP2/A3/A3.java
import java.io.*;
import java.lang.*;
import java.util.*;


class GreedySearchAlgorithms {

    static void selectionSort(int[] A) {
        int[] U = A.clone();
        int n = A.length;
        for (int i = 0; i < n - 1; i++) {
            int min_idx = i;
            for (int j = i + 1; j < n; j++) {
                if (A[j] < A[min_idx]) {
                    min_idx = j;
                }
            }
            int tmp = A[i];
            A[i] = A[min_idx];
            A[min_idx] = tmp;
        }

        System.out.printf("Selection Sort:\nUnsorted array: %s\nSorted array: %s", Arrays.toString(U), Arrays.toString(A));
    }

    static void jobScheduling(ArrayList<Job> arr) {
        int n = arr.size();
        Collections.sort(arr, (a, b) -> {
            return a.deadline - b.deadline;
        });

        ArrayList<Job> result = new ArrayList<>();
        PriorityQueue<Job> maxHeap = new PriorityQueue<>((a, b) -> {
            return b.profit - a.profit;
        });

        for (int i = n - 1; i > -1; i--) {
            int slot_available;
            if (i == 0)
                slot_available = arr.get(i).deadline;
            else
                slot_available = arr.get(i).deadline - arr.get(i - 1).deadline;

            maxHeap.add(arr.get(i));

            while (slot_available > 0 && maxHeap.size() > 0) {
                Job job = maxHeap.remove();
                slot_available--;
                result.add(job);
            }
        }

        Collections.sort(result, (a, b) -> {
            return a.deadline - b.deadline;
        });

        System.out.print("\n\n\nJob Scheduling Problem:\nFollowing is maximum profit sequence of jobs: \n");
        for (Job job : result) {
            System.out.printf("[%s, %d, %d] -> ", job.job_id, job.deadline, + job.profit);
        }
        System.out.print("Finish");
    }

    static int minDistance(int[] dist, boolean[] visited) {
        int min = Integer.MAX_VALUE;
        int min_index = -1;

        for (int i = 0; i < dist.length; i++) {
            if (visited[i] == false && dist[i] <= min) {
                min = dist[i];
                min_index = i;
            }
        }
        return min_index;
    }

    static void dijkstra(int[][] graph, int src, int dest) {
        int n = graph.length;
        int[] dist = new int[n];
        boolean[] visited = new boolean[n];
        HashMap<Integer, ArrayList<Integer>> parent = new HashMap<>();
        List<Integer> path = new ArrayList<>();
        path.add(dest);

        for (int i = 0; i < n; i++) {
            dist[i] = Integer.MAX_VALUE;
            visited[i] = false;
        }

        dist[src] = 0;
        parent.put(src, new ArrayList<>());

        for (int i = 0; i < n - 1; i++) {
            int u = minDistance(dist, visited);
            visited[u] = true;

            for (int v = 0; v < n; v++) {
                if (
                    !visited[v] && graph[u][v] != 0 &&
                    dist[u] != Integer.MAX_VALUE &&
                    dist[u] + graph[u][v] < dist[v]
                ) {
                    dist[v] = dist[u] + graph[u][v];

                    if (!parent.containsKey(v))
                        parent.put(v, new ArrayList<Integer>());
                    parent.get(v).add(u);
                }
            }
        }

        int key = dest;
        while (parent.get(key).size() > 0) {
            int elem = parent.get(key).get(parent.get(key).size()-1);;
            path.add(elem);
            key = elem;
        }
        Collections.reverse(path);

        System.out.printf("\n\n\nDijkstra Single-Source Shortest Path::\nPath: %s\nMinimum Cost: %d", path.toString(), dist[dest]);
    }

    public static void main(String args[]) {

        selectionSort(new int[]{64, 25, 12, 22, 11});

        jobScheduling(new ArrayList<Job>(Arrays.asList(
            new Job('A', 2, 100),
            new Job('B', 1, 19),
            new Job('C', 2, 27),
            new Job('D', 1, 25),
            new Job('E', 3, 15)
        )));

		// A2.jpg
		int graph[][] = new int[][] {
            { 0, 4, 0, 0, 0, 0, 0, 8, 0},
            { 4, 0, 8, 0, 0, 0, 0,11, 0},
            { 0, 8, 0, 7, 0, 4, 0, 0, 2},
            { 0, 0, 7, 0, 9,14, 0, 0, 0},
            { 0, 0, 0, 9, 0,10, 0, 0, 0},
            { 0, 0, 4,14,10, 0, 2, 0, 0},
            { 0, 0, 0, 0, 0, 2, 0, 1, 6},
            { 8,11, 0, 0, 0, 0, 1, 0, 7},
            { 0, 0, 2, 0, 0, 0, 6, 7, 0}
        };

		PrimsMST primsMST = new PrimsMST(9, graph);
		primsMST.primMST();

		KruskalsMST kruskalsMST = new KruskalsMST(9, 14, graph);
		kruskalsMST.KruskalMST();

        dijkstra(graph, 0, 4);
    }
}


class Job {
    char job_id;
    int deadline, profit;
    Job(char job_id, int deadline, int profit) {
        this.deadline = deadline;
        this.job_id = job_id;
        this.profit = profit;
    }
}


class PrimsMST {
	private int V, graph[][];

    PrimsMST(int V, int graph[][]) {
        this.V = V;
        this.graph = graph;
    }

	int minKey(int key[], Boolean mstSet[]) {
		int min = Integer.MAX_VALUE, min_index = -1;
		for (int v = 0; v < V; v++)
			if (mstSet[v] == false && key[v] < min) {
				min = key[v];
				min_index = v;
			}
		return min_index;
	}

	void primMST() {
		int parent[] = new int[V];
		int key[] = new int[V];
		Boolean mstSet[] = new Boolean[V];

		for (int i = 0; i < V; i++) {
			key[i] = Integer.MAX_VALUE;
			mstSet[i] = false;
		}

		key[0] = 0;
		parent[0] = -1;

		for (int count = 0; count < V - 1; count++) {
			int u = minKey(key, mstSet);
			mstSet[u] = true;

			for (int v = 0; v < V; v++)
				if (graph[u][v] != 0 && mstSet[v] == false && graph[u][v] < key[v]) {
					parent[v] = u;
					key[v] = graph[u][v];
				}
		}

		System.out.println("\n\n\nPrim’s Minimum Spanning Tree:\nEdge \tWeight");
        int minimumCost = 0;
		for (int i = 1; i < V; i++){
			System.out.printf("%d -- %d == %d\n", parent[i], i, graph[i][parent[i]]);
            minimumCost += graph[i][parent[i]];
        }
        System.out.printf("Minimum Cost: %d", minimumCost);
	}
}


class KruskalsMST {
	class Edge implements Comparable<Edge> {
		int src, dest, weight;
		public int compareTo(Edge compareEdge) {
			return this.weight - compareEdge.weight;
		}
	};

	class subset {
		int parent, rank;
	};

	private int V, E;
	private Edge edge[];

	KruskalsMST(int v, int e, int graph[][]) {
		this.V = v;
		this.E = e;
		this.edge = new Edge[E];
        int i = -1;
		for (int x = 0; x < v; x++)
            for (int y = x; y < v; y++)
                if (graph[x][y] != 0) {
                    edge[++i] = new Edge();
                    edge[i].src = x;
                    edge[i].dest = y;
                    edge[i].weight = graph[x][y];
                }
	}

	int find(subset subsets[], int i) {
		if (subsets[i].parent != i)
			subsets[i].parent = find(subsets, subsets[i].parent);
		return subsets[i].parent;
	}

	void Union(subset subsets[], int x, int y) {
		int xroot = find(subsets, x);
		int yroot = find(subsets, y);

		if (subsets[xroot].rank < subsets[yroot].rank)
			subsets[xroot].parent = yroot;
		else if (subsets[xroot].rank > subsets[yroot].rank)
			subsets[yroot].parent = xroot;
		else {
			subsets[yroot].parent = xroot;
			subsets[xroot].rank++;
		}
	}

	void KruskalMST() {
		Edge result[] = new Edge[V];
		int e = 0, i = 0;

		for (i = 0; i < V; ++i)
			result[i] = new Edge();
		Arrays.sort(edge);

		subset subsets[] = new subset[V];
		for (i = 0; i < V; ++i)
			subsets[i] = new subset();

		for (int v = 0; v < V; ++v) {
			subsets[v].parent = v;
			subsets[v].rank = 0;
		}

		i = 0;

		while (e < V - 1) {
			Edge next_edge = edge[i++];

			int x = find(subsets, next_edge.src);
			int y = find(subsets, next_edge.dest);

			if (x != y) {
				result[e++] = next_edge;
				Union(subsets, x, y);
			}
		}

		System.out.println("\n\n\nKruskal’s Minimum Spanning Tree:\nEdge \tWeight");
		int minimumCost = 0;
		for (i = 0; i < e; ++i) {
			System.out.printf("%d -- %d == %d\n", result[i].src, result[i].dest, result[i].weight);
			minimumCost += result[i].weight;
		}
        System.out.printf("Minimum Cost: %d", minimumCost);
	}
}
